### PR TITLE
layer_store: Use CreateReadWrite() for -init layer instead of Create()

### PR DESCRIPTION
**Pull request by rhvgoyal**
init layer is read/write layer and not read only layer. Following commit
introduced new graph driver method CreateReadWrite.

ef5bfad Adding readOnly parameter to graphdriver Create method

So far only windows seem to be differentiating between above two methods.
Making this change to make sure -init layer calls right method so that
we don't have surprises in future.

Signed-off-by: Vivek Goyal vgoyal@redhat.com

### CODE REVIEW

**Comment by rhvgoyal** [comment]
CC @swernli

**Comment by swernli** [comment]
@rhvgoyal Unfortunately, I don't think this is a change we should take right now.  While this makes perfect logical sense, it has a big perf impact on Windows (creating a read/write layer is slow on Windows), and the init layer is completely unused by Windows containers (`initFunc` comes from `daemon.setupInitLayer` which is a no-op).

I think we should take this fix only after updating docker to stop creating an init layer on Windows, otherwise the perf impact is too high.

**Comment by rhvgoyal** [comment]
@swernli If windows does not need init layer, then how about the new patch I pushed. This patch stops creation of init layer on windows.

**Comment by swernli** [comment]
@rhvgoyal Excellent, yes this is exactly what I was hoping for.  Thanks!

LGTM.

**Comment by rhvgoyal** [comment]
cc @dmcgowan @crosbymichael

**Comment by dmcgowan** [comment]
Could be a smaller change to instead have a `func (daemon *Daemon) getLayerInit() func(string) error` with different implementation for windows and unix, but that seems quite nitty.

LGTM

**Comment by rhvgoyal** [comment]
@dmcgowan I like your idea of returning function pointer. Pushed a new patch. PTAL.

**Comment by swernli** [comment]
Even better!

**Comment by dmcgowan** [comment]
LGTM

**Comment by rhvgoyal** [comment]
This windowsRS1 error, I doubt it is because of this PR.

**Comment by swernli** [comment]
@rhvgoyal A fix for that issue was actually just merged in: https://github.com/docker/docker/pull/26764
@darrenstahlmsft can confirm that this is the same, or you could rebase and push to retry CI with the latest.

**Comment by darrenstahlmsft** [comment]
Yes, the windowsRS1 failure is fixed by that PR. I've restarted the windowsRS1 CI to pick up the fix.

File: daemon/create.go [hunk 1]
```diff
@@ -201,7 +201,9 @@ func (daemon *Daemon) setRWLayer(container *container.Container) error {
 		}
 		layerID = img.RootFS.ChainID()
 	}
-	rwLayer, err := daemon.layerStore.CreateRWLayer(container.ID, layerID, container.MountLabel, daemon.setupInitLayer, container.HostConfig.StorageOpt)
+
+	rwLayer, err := daemon.layerStore.CreateRWLayer(container.ID, layerID, container.MountLabel, daemon.getLayerInit(), container.HostConfig.StorageOpt)
+
```
Comment (thaJeztah) [inline daemon/create.go]: _really_ nitty, but unrelated whitespace added?

File: daemon/daemon_unix.go [hunk 2]
```diff
@@ -779,6 +779,10 @@ func initBridgeDriver(controller libnetwork.NetworkController, config *Config) e
 	return nil
 }
 
+func (daemon *Daemon) getLayerInit() func(string) error {
+	return daemon.setupInitLayer
```
Comment (thaJeztah) [inline daemon/daemon_unix.go]: Should probably be added to `daemon_solaris.go` as well?

**Comment by crosbymichael** [comment]
LGTM

File: daemon/daemon_unix.go (hunk 2 above)
Comment (rhvgoyal) [inline daemon/daemon_unix.go]: Hmm.., I had not noticed daemon_solaris.go. Looks like we should put it in there as well. Looks like CI does not builds for solaris? Otherwise build should have failed.

Anyway, I will update patch to take care of solaris as well.

File: daemon/create.go (hunk 1 above)
Comment (rhvgoyal) [inline daemon/create.go]: You mean, extra line added? I can get rid of it. Not a problem.

File: daemon/daemon_unix.go (hunk 2 above)
Comment (thaJeztah) [inline daemon/daemon_unix.go]: It's still being worked on, so no CI yet, thanks for doing the follow up!

File: daemon/create.go (hunk 1 above)
Comment (thaJeztah) [inline daemon/create.go]: It's merged now, so no real need to do so (was just trying to keep the diff small - easier to check history :smile:)
